registry/api_test.go

```diff
@@ -120,29 +121,59 @@ func TestLayerAPI(t *testing.T) {
 	checkResponse(t, "checking head on non-existent layer", resp, http.StatusNotFound)
 
 	// ------------------------------------------
-	// Upload a layer
-	layerUploadURL, err := builder.BuildBlobUploadURL(imageName)
+	// Start an upload and cancel
+	uploadURLBase := startPushLayer(t, builder, imageName)
+
+	req, err := http.NewRequest("DELETE", uploadURLBase, nil)
 	if err != nil {
-		t.Fatalf("error building upload url: %v", err)
+		t.Fatalf("unexpected error creating delete request: %v", err)
 	}
 
-	resp, err = http.Post(layerUploadURL, "", nil)
+	resp, err = http.DefaultClient.Do(req)
 	if err != nil {
-		t.Fatalf("error starting layer upload: %v", err)
+		t.Fatalf("unexpected error sending delete request: %v", err)
 	}
 
-	checkResponse(t, "starting layer upload", resp, http.StatusAccepted)
-	checkHeaders(t, resp, http.Header{
-		"Location":       []string{"*"},
-		"Content-Length": []string{"0"},
-	})
+	checkResponse(t, "deleting upload", resp, http.StatusNoContent)
+
+	// A status check should result in 404
+	resp, err = http.Get(uploadURLBase)
+	if err != nil {
+		t.Fatalf("unexpected error getting upload status: %v", err)
+	}
+	checkResponse(t, "status of deleted upload", resp, http.StatusNotFound)
+
+	// -----------------------------------------
+	// Do layer push with an empty body
+	uploadURLBase = startPushLayer(t, builder, imageName)
+	resp, err = doPushLayer(t, builder, imageName, layerDigest, uploadURLBase, bytes.NewReader([]byte{}))
+	if err != nil {
+		t.Fatalf("unexpected error doing bad layer push: %v", err)
+	}
 
+	checkResponse(t, "bad layer push", resp, http.StatusBadRequest)
+	checkBodyHasErrorCodes(t, "bad layer push", resp, v2.ErrorCodeBlobUploadInvalid)
+
+	// -----------------------------------------
+	// Do layer push with an invalid body
+
```

Comment:
We are probably missing two cases here:

1. Empty tar file (zero-length) with the correct digest should be accepted.
2. Tar file with block of 1024 zero-valued bytes (a valid tar) with correct digest should be accepted.

Both of these cases result in a 400 error with this changeset.